### ISSUE TITLE
Multi-Process Single-GPU is bad

### ISSUE THREAD

**Issue by tstandley** [pytorch/pytorch#37444]
I'm sorry to do this, but since your release states that you plan to deprecate Single-Process Multi-GPU, I have to talk about all the issues I have with MPSG.

First, this mode takes up a CPU thread per GPU you have. I'm sure that the developers of pytorch are used to working with servers that have tens of CPU cores and the number of cores is much much greater than the number of GPUs, but students like me who are trying to stretch their compute budgets don't necessarily have this luxury. For example, I have a workstation with 4 GPU's and 8 cores. Another with 2 GPU's and 4 cores. Those extra threads are needed for data loading, let alone anything else I want to do with the machine.

Second, killing jobs is hard and often results in zombie processes that have to be tracked down and killed manually. CTRL-c should kill all processes and the user shouldn't have to do anything else. This is how it works with DataParallel. On a related note, these extra processes often output stuff to the console in duplicate, which really clutters output. Sometimes even after the head is killed.

Third, the mode is error prone when it comes to data loading. We're supposed to use DistributedSampler to ensure that each node loads a different partition of the data. It's not easy to verify that it is working and not creating duplicates, and I'm often trying to do something just a little fancier than what DistributedSampler was designed to do. Failure in this case is silent. You just get worse results.

Finally, I just don't like the API of it. It feels like there's all this code clutter "if args.distributed" everywhere and you have to call "reduce_tensor" for each metric you compute. If you forget, everything still "works" your results are just inaccurate. You also have to launch your program with "python -m torch.distributed.launch --nproc_per_node=xx" which is esoteric and clutters your console. 

As a result, I find myself using DataParallel almost all the time, even though I know it's slower. It's usually not that much slower though. I hope to be able to keep it.

What I really hope for is more support for it. Why is it slower? Can that speed gap be closed somewhat? It uses more GPU memory too, especially on GPU 0. Trying to use the new "Channels Last" with it results in no speedup, but if I test on a single GPU, sometimes the speedup is significant.

Thanks! I really appreciate PyTorch on the whole. It's an invaluable tool for students like me.

**Comment by ngimel** [comment]
DataParallel is usually as fast (or as slow) as single-process multi-GPU. Those extra threads for multi-process single-GPU are used not for frivolous reason, but because single thread is usually not fast enough to feed multiple GPUs. 
You points about API clunkiness and hard-to-kill jobs are valid, we need to make it easier.

**Comment by tstandley** [comment]
Thanks for the quick response! Does the CPU thread only bottleneck the GPUs sometimes? For example if your batch size/network is small? When else? What work is that thread doing except for telling the GPU's to do X computation and then Y computation, etc.? Does it take longer to tell multiple GPU's that?

Either way, it'd be nice to only pay the cost of that extra thread if it was necessary. I'm using mostly consumer desktop CPUs, so threads are fast, but you don't get as many of them.

Would it be possible to design the system to use a different number of threads than GPUs. For example 2 threads for 4 GPUs?

**Comment by mrshenli** [comment]
@tstandley just to clarify this:

We are **not** going to deprecate **`DataParallel`**.

This is not only for the above reasons you mentioned, but also because `DataParallel` only requires minimum change in the application code to enable parallelism. So this is a better entry point than `DistributedDataParallel` to get started with data parallel training. 

What we are planning to deprecate is the `DataParallel`-like mode in `DistributedDataParallel` (DDP). So again, `DataParallel` will stay **intact** and the change only goes into `DistributedDataParallel`.

The motivations for retiring SPMG is as @ngimel mentioned above. On a single machine, DDP in SPMG mode is very similar to `DataParallel`, and the main difference is whether it replicate model in every iteration or in the constructor. So there is no reason to keep two mostly the same but subtly different implementations of SPMG, as it would only introduce more code maintenance overhead and make users confused (I guess is this is why we are discussing this here? :) )

Compared to `DataParallel` which aims for simplicity, `DistributedDataParallel` requires more setup to use, but aims for scalability and performance. For applications that care a lot of performance, it is usually OK to spend a few more lines of code to setup process group, data loader etc. So I guess the first three points you mentioned are fine in this case?

> Finally, I just don't like the API of it. It feels like there's all this code clutter "if args.distributed" everywhere and you have to call "reduce_tensor" for each metric you compute. If you forget, everything still "works" your results are just inaccurate. You also have to launch your program with "python -m torch.distributed.launch --nproc_per_node=xx" which is esoteric and clutters your console.

For the API, you don't have to use the launch script. That is just one tool to help people who need it. If you are using it on a single machine, there are simpler ways to launch. See [this example](https://pytorch.org/docs/master/notes/ddp.html).


BTW, where did you saw that we are going to deprecate it in the release note or somewhere else?

**Comment by mrshenli** [comment]
@tstandley 

> Does the CPU thread only bottleneck the GPUs sometimes? For example if your batch size/network is small? When else? What work is that thread doing except for telling the GPU's to do X computation and then Y computation, etc.? Does it take longer to tell multiple GPU's that?

The slowness of `DataParallel` comes from several places.

1. GIL contention across multiple threads. 
2. Replicating model in every iteration
3. Additional overhead caused by scattering inputs and gathering outputs.

We probably to get rid of 2 if we do that in constructor. But 1 and 3 will stay.

**Comment by mrshenli** [comment]
> Trying to use the new "Channels Last" with it results in no speedup, but if I test on a single GPU, sometimes the speedup is significant.

This probably worth a dedicated issue to track. Could you please create a new issue with a minimum repro for this? We will investigate why there is no speedup.

**Comment by ngimel** [comment]
Typically CPU threads are the bottleneck in small batch/network situation as you note. But as GPUs become faster the definition of "small" shifts also, and more workloads fall into this category.
If the threads are not pinned, you are not really paying the cost of extra threads if not necessary, there's no to little harm in oversubscribing if most of the time threads are idle anyway.  ymmv of course, and there are pathological cases where oversubscription would lead to dramatic slowdowns. 
It's possible to design a system like this, and this is what multi-process multi-gpu ddp used to be, but in the end it just becomes too cumbersome to support. We should be able to improve MPSG UX with ddp, and "not enough CPU threads" should not be a dealbreaker for that, unless you really have a lot of work for cpu threads and they can't keep up, but in this case DataParallel won't save you either, it's just not enough resources.

**Comment by tstandley** [comment]
@mrshenli Thank you for your responses! Yes. I misinterpreted the release notes for pytorch 1.5. I'm quite happy that I'll still have the option for DataParallel and SPMG. Any support for that method of training would be much appreciated. Right now, I get the feeling that it's something you'd rather your users not do. 

(3) probably isn't much overhead, right? Anything you could do to fix (2) would be awesome! (1) is a headache for all python users.

And to be clear, I totally see the benefit of MPSG, it's definitely worth it to get more speed in many situations. I'll go quite a bit out of the way to squeeze more performance out of my hardware.

Is there anything to be done to address my third issue with MPSG--data loading? I want to be sure everything is right, even if I want to do something like have each epoch be only 20% of my data, but only repeat a datum once every 5 epochs. Correctness is simply more important than speed.

**Comment by mrshenli** [comment]
cc @vincentqb for data loader discussion :)

**Comment by tstandley** [comment]
Also, it'd be great if batch norm syncing worked with DataParallel.

**Comment by vincentqb** [comment]
> Is there anything to be done to address my third issue with MPSG--data loading? I want to be sure everything is right, even if I want to do something like have each epoch be only 20% of my data, but only repeat a datum once every 5 epochs. Correctness is simply more important than speed.

You mean something like using a [subset](https://pytorch.org/docs/stable/data.html#torch.utils.data.Subset) of your dataset when debugging?
